File: baekjoon/old/5052_phoneBook.py
success = "YES"

class TreeNode:
  def __init__(self, phoneNumber):
    self.next = [ 0 for _ in range(11)]
    if(len(phoneNumber) == 0):
        self.next[-1] = 1
        return
    number = int(phoneNumber[0])
    restNumbers = phoneNumber[1:]
    self.next[number] = TreeNode(restNumbers)

  def insert(self, phoneNumber):
    global success
    if(self.next[-1]):
        success = "NO"
        return
    if(len(phoneNumber) == 0):
        success = "NO"
        return
    number = int(phoneNumber[0])
    restNumbers = phoneNumber[1:]
    if(self.next[number]):
        self.next[number].insert(restNumbers)
    else:
        self.next[number] = TreeNode(restNumbers)
  
def solution():
    global success
    success = "YES"
    N = int(input())
    phoneNumberBook = []

    for _ in range(N):
        phoneNumber = input()
        phoneNumberBook.append(phoneNumber)
    
    root = TreeNode(phoneNumberBook[0])
    for phoneNumber in phoneNumberBook[1:]:
        root.insert(phoneNumber)

T = int(input())
for i in range(T):
    solution()
    print(success)

File: baekjoon/old/5052_phoneBook.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int test_case();
long long solution();
int T;

class TreeNode{
    TreeNode* next[10];
    bool isLeaf;
public:
    TreeNode(string &phoneNumber, int depth) {
        for (int i = 0; i < 10; i++) {
            next[i] = NULL;
        }
        if(phoneNumber.size() == depth) {
            isLeaf = true;
            return;
        }
        isLeaf = false;
        int number = phoneNumber[depth] - 0x30;
        next[number] = new TreeNode(phoneNumber, depth+1);
    }

    bool insert(string &phoneNumber, int depth){
        if(isLeaf || phoneNumber.size() == depth){
            return false;
        }
        int number = phoneNumber[depth] - 0x30;
        if(next[number]) {
            return next[number]->insert(phoneNumber, depth+1);
        }
        else {
            next[number] = new TreeNode(phoneNumber, depth+1);
            return true;
        }
    }
    // delete
};

int main()
{
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    test_case();
    return 0;
}

long long solution()
{
    // input
    int N;
    cin >> N;
    vector<string> phoneNumberBook;
    phoneNumberBook.resize(N);
    for(int i = 0; i < N; i++)
    {
        cin >> phoneNumberBook[i];
    }
    // solution
    TreeNode *root;
    root = new TreeNode(phoneNumberBook[0], 0);
    for(int i = 1; i < N; i++){
        bool success = root->insert(phoneNumberBook[i], 0);
        if(!success) {
            cout << "NO\n";
            return 0;
        }
    }
    cout << "YES\n";
    return 0;
}

int test_case()
{
    cin >> T;
    for(int i = 0; i < T; i++)
    {
        solution();
    }
    return 0;
}
